### ISSUE TITLE
amplify mock Error: Cannot find ref to userpoolId

### ISSUE THREAD

**Issue by mooser** [aws-amplify/amplify-cli#2453]
I am trying to get "amplify mock" to work, but I'm getting 2 errors: 

> Failed to parse Lambda function cloudformation in
> amplify/backend/function/lambdaHandler/lambdaHandler-cloudformation-template.json
> 
> Error: Could not find ref for "mycustomname77b3f30UserPoolId"
> 
> Also in the lambda that was auto generated by "amplify add auth" "mycustomnamePostConfirmation"
> 
> mycustomnamePostConfirmation-cloudformation-template.json
> Error: Could not find ref for "GROUP"

There's a function in my project called "lambdaHandler"  (amplify add function)
API: yes 
Auth: yes (cognito)

Any ideas at all on what must be done here to get a local mock environment working?

**Comment by mooser** [comment]
I was able to recreate the issue...  Now I just gotta find some documentation on how to get the ref resolved...

Create-react-app
amplify init
amplify add auth (choose federated with cognito)
amplify add api  (graphql)
amplify add function (called uraHandler)
  - tell it to connect to auth and api

Then in the schema I added the resolver directive to use the lambda.
@function(name: "uraHandler-${env}")

backend-config.json
```
{
	"auth": {
		"testapp31eeffd2": {
			"service": "Cognito",
			"providerPlugin": "awscloudformation",
			"dependsOn": []
		}
	},
	"api": {
		"testapi": {
			"service": "AppSync",
			"providerPlugin": "awscloudformation",
			"output": {
				"authConfig": {
					"additionalAuthenticationProviders": [],
					"defaultAuthentication": {
						"authenticationType": "AMAZON_COGNITO_USER_POOLS",
						"userPoolConfig": {
							"userPoolId": "authtestapp31eeffd2"
						}
					}
				}
			}
		}
	},
	"function": {
		"uraHandler": {
			"service": "Lambda",
			"providerPlugin": "awscloudformation",
			"build": true,
			"dependsOn": [
				{
					"category": "auth",
					"resourceName": "testapp31eeffd2",
					"attributes": [
						"UserPoolId"
					]
				},
				{
					"category": "api",
					"resourceName": "testapi",
					"attributes": [
						"GraphQLAPIIdOutput",
						"GraphQLAPIEndpointOutput"
					]
				}
			]
		}
	}
}
```

uraHandler-cloudformation-template.json
```
{
    "AWSTemplateFormatVersion": "2010-09-09",
    "Description": "Lambda resource stack creation using Amplify CLI",
    "Parameters": {
        "env": {
            "Type": "String"
        },
    
    
    
        "authtestapp31eeffd2UserPoolId": {
            "Type": "String",
            "Default": "authtestapp31eeffd2UserPoolId"
            },
        
        
    
        "apitestapiGraphQLAPIIdOutput": {
            "Type": "String",
            "Default": "apitestapiGraphQLAPIIdOutput"
            },
        
        "apitestapiGraphQLAPIEndpointOutput": {
            "Type": "String",
            "Default": "apitestapiGraphQLAPIEndpointOutput"
            }
        
        
    
    },
    "Conditions": {
        "ShouldNotCreateEnvResources": {
            "Fn::Equals": [
                {
                    "Ref": "env"
                },
                "NONE"
            ]
        }
    },
    "Resources": {
        "LambdaFunction": {
          "Type": "AWS::Lambda::Function",
          "Metadata": {
            "aws:asset:path": "./src",
            "aws:asset:property": "Code"
          },
          "Properties": {
            "Handler": "index.handler",
            "FunctionName": {
                "Fn::If": [
                    "ShouldNotCreateEnvResources",
                    "uraHandler", 
                    {

                        "Fn::Join": [
                            "",
                            [
                                "uraHandler",
                                "-",
                                {
                                    "Ref": "env"
                                }
                            ]
                        ]
                    }      
                ]
            },
            "Environment": {
                "Variables" : {
                    "ENV": {
                        "Ref": "env"
                    },
                    "REGION": { 
                        "Ref": "AWS::Region"
                    }
                    ,"AUTH_TESTAPP31EEFFD2_USERPOOLID": {"Ref": "authtestapp31eeffd2UserPoolId"},"API_TESTAPI_GRAPHQLAPIIDOUTPUT": {"Ref": "apitestapiGraphQLAPIIdOutput"},"API_TESTAPI_GRAPHQLAPIENDPOINTOUTPUT": {"Ref": "apitestapiGraphQLAPIEndpointOutput"} 
                }
            },
            "Role": { "Fn::GetAtt" : ["LambdaExecutionRole", "Arn"] },
            "Runtime": "nodejs10.x",
            "Timeout": "25"
          }
        },
        "LambdaExecutionRole": {
            "Type": "AWS::IAM::Role",
            "Properties": {
                "RoleName": {
                    "Fn::If": [
                        "ShouldNotCreateEnvResources",
                        "testappLambdaRolea65e2ff3", 
                        {

                            "Fn::Join": [
                                "",
                                [
                                    "testappLambdaRolea65e2ff3",
                                    "-",
                                    {
                                        "Ref": "env"
                                    }
                                ]
                            ]
                        } 
                    ]
                },
                "AssumeRolePolicyDocument": {
                    "Version": "2012-10-17",
                    "Statement": [
                        {
                            "Effect": "Allow",
                            "Principal": {
                                "Service": [
                                    "lambda.amazonaws.com"
                                ]
                            },
                            "Action": [
                                "sts:AssumeRole"
                            ]
                        }
                    ]
                }
            }
        }
        ,"lambdaexecutionpolicy": {
            "DependsOn": ["LambdaExecutionRole"],
            "Type": "AWS::IAM::Policy",
            "Properties": {
                "PolicyName": "lambda-execution-policy",
                "Roles": [{ "Ref": "LambdaExecutionRole" }],
                "PolicyDocument": {
                    "Version": "2012-10-17",
                    "Statement": [
                        {
                            "Effect": "Allow",
                            "Action":["logs:CreateLogGroup",
                            "logs:CreateLogStream",
                            "logs:PutLogEvents"],
                            "Resource": { "Fn::Sub" : [ "arn:aws:logs:${region}:${account}:log-group:/aws/lambda/${lambda}:log-stream:*", { "region": {"Ref": "AWS::Region"},  "account": {"Ref": "AWS::AccountId"}, "lambda": {"Ref": "LambdaFunction"}} ]}
                        }
                    ]
                }
            }
        }
        ,"AmplifyResourcesPolicy": {
            "DependsOn": ["LambdaExecutionRole"],
            "Type": "AWS::IAM::Policy",
            "Properties": {
                "PolicyName": "amplify-lambda-execution-policy",
                "Roles": [{ "Ref": "LambdaExecutionRole" }],
                "PolicyDocument": {
                    "Version": "2012-10-17",
                    "Statement": [{"Effect":"Allow","Action":["cognito-idp:ConfirmSignUp","cognito-idp:AdminCreateUser","cognito-idp:CreateUserImportJob","cognito-idp:AdminSetUserSettings","cognito-idp:AdminLinkProviderForUser","cognito-idp:CreateIdentityProvider","cognito-idp:AdminConfirmSignUp","cognito-idp:AdminDisableUser","cognito-idp:AdminRemoveUserFromGroup","cognito-idp:SetUserMFAPreference","cognito-idp:SetUICustomization","cognito-idp:SignUp","cognito-idp:VerifyUserAttribute","cognito-idp:SetRiskConfiguration","cognito-idp:StartUserImportJob","cognito-idp:AdminSetUserPassword","cognito-idp:AssociateSoftwareToken","cognito-idp:CreateResourceServer","cognito-idp:RespondToAuthChallenge","cognito-idp:CreateUserPoolClient","cognito-idp:AdminUserGlobalSignOut","cognito-idp:GlobalSignOut","cognito-idp:AddCustomAttributes","cognito-idp:CreateGroup","cognito-idp:CreateUserPool","cognito-idp:AdminForgetDevice","cognito-idp:AdminAddUserToGroup","cognito-idp:AdminRespondToAuthChallenge","cognito-idp:ForgetDevice","cognito-idp:CreateUserPoolDomain","cognito-idp:AdminEnableUser","cognito-idp:AdminUpdateDeviceStatus","cognito-idp:StopUserImportJob","cognito-idp:InitiateAuth","cognito-idp:AdminInitiateAuth","cognito-idp:AdminSetUserMFAPreference","cognito-idp:ConfirmForgotPassword","cognito-idp:SetUserSettings","cognito-idp:VerifySoftwareToken","cognito-idp:AdminDisableProviderForUser","cognito-idp:SetUserPoolMfaConfig","cognito-idp:ChangePassword","cognito-idp:ConfirmDevice","cognito-idp:AdminResetUserPassword","cognito-idp:ResendConfirmationCode","cognito-identity:Describe*","cognito-identity:Get*","cognito-identity:List*","cognito-idp:Describe*","cognito-idp:AdminGetDevice","cognito-idp:AdminGetUser","cognito-idp:AdminList*","cognito-idp:List*","cognito-sync:Describe*","cognito-sync:Get*","cognito-sync:List*","iam:ListOpenIdConnectProviders","iam:ListRoles","sns:ListPlatformApplications","cognito-idp:ForgotPassword","cognito-idp:UpdateAuthEventFeedback","cognito-idp:UpdateResourceServer","cognito-idp:UpdateUserPoolClient","cognito-idp:AdminUpdateUserAttributes","cognito-idp:UpdateUserAttributes","cognito-idp:UpdateUserPoolDomain","cognito-idp:UpdateIdentityProvider","cognito-idp:UpdateGroup","cognito-idp:AdminUpdateAuthEventFeedback","cognito-idp:UpdateDeviceStatus","cognito-idp:UpdateUserPool","cognito-idp:DeleteUserPoolDomain","cognito-idp:DeleteResourceServer","cognito-idp:DeleteGroup","cognito-idp:AdminDeleteUserAttributes","cognito-idp:DeleteUserPoolClient","cognito-idp:DeleteUserAttributes","cognito-idp:DeleteUserPool","cognito-idp:AdminDeleteUser","cognito-idp:DeleteIdentityProvider","cognito-idp:DeleteUser"],"Resource":[{"Fn::Join":["",["arn:aws:cognito-idp:",{"Ref":"AWS::Region"},":",{"Ref":"AWS::AccountId"},":userpool/",{"Ref":"authtestapp31eeffd2UserPoolId"}]]}]},{"Effect":"Allow","Action":["appsync:Create*","appsync:StartSchemaCreation","appsync:GraphQL","appsync:Get*","appsync:List*","appsync:Update*","appsync:Delete*"],"Resource":[{"Fn::Join":["",["arn:aws:appsync:",{"Ref":"AWS::Region"},":",{"Ref":"AWS::AccountId"},":apis/",{"Ref":"apitestapiGraphQLAPIIdOutput"},"/*"]]}]}]
                }
            }
        }
        
    },
    "Outputs": {
        "Name": {
            "Value": {
                "Ref": "LambdaFunction"
            }
        },
        "Arn": {
            "Value": {"Fn::GetAtt": ["LambdaFunction", "Arn"]}
        },
        "Region": {
            "Value": {
                "Ref": "AWS::Region"
            }
        },
        "LambdaExecutionRole": {
            "Value": {
                "Ref": "LambdaExecutionRole"
            }
        }
    }
}
```

**Comment by mooser** [comment]
I managed to hack some configs and get amplify mock working for the first time.  WooHoo I am unblocked.  Just note: by default amplify sets up a project with a broken amplify mock if you add cognito + lambda resolver that depends on it.  

I am in the process of trying to get source level debugging working for lambda functions locally with amplify mock.  I am hoping I can attach to the local graphql endpoint and hit breakpoints in my lambda with vscode...  We'll see how it goes.

**Comment by warrenmcquinn** [comment]
I'm having the same issue.

```
Failed to parse Lambda function cloudformation in amplify/backend/function/customResolvers/customResolvers-cloudformation-template.json

Error: Could not find ref for "...UserPoolId"
Failed to start API Mock endpoint Error: Lambda function functionName does not exist in your project.
Please run amplify add function
```

even though the function `functionName` is in my project / environment.

I'm unable to use `amplify mock api` as a result.

**Comment by zaklampert** [comment]
@mooser what config hacks worked for you?

**Comment by mooser** [comment]
@zaklampert You can specify the missing refs in function-parameters.json

So if you are missing ref for "GROUP" your function-parameters.json should look like: 
`{ "GROUP": "group-name"} `

**Comment by rgilks** [comment]
I also have a similar problem, though not to do with the userPoolId but another ref from the cloud-formation template.

`
Failed to parse Lambda function cloudformation in amplify/backend/function/orderProcessor/orderProcessor-cloudformation-template.json

Error: Could not find ref for "apiprediraGraphQLAPIIdOutput"
Failed to start API Mock endpoint Error: Lambda function orderProcessor does not exist in your project. 
Please run amplify add function
`

The function definitely exists. Everything is fine other than `amplify mock` no longer works. It did about a month ago though.

**Comment by theslyone** [comment]
@rgilks you need to add "apiprediraGraphQLAPIIdOutput" as a variable to the function-parameters.json and point it to your local graphql instance 
e.g `"apifluxGraphQLAPIEndpointOutput": "http://localhost:20002/graphql"

**Comment by imgntn** [comment]
same problem.  all i did was a single function directive to my schema and `amplify mock api` no longer works.  

```
Failed to parse Lambda function cloudformation in [redacted]\amplify\backend\function\createRenderJob\createRenderJob-cloudformation-template.json

Error: Could not find ref for "apiGQLAPIGraphQLAPIIdOutput"
Failed to parse Lambda function cloudformation in [redacted]\amplify\backend\function\handlePayment\handlePayment-cloudformation-template.json

Error: Could not find ref for "apiGQLAPIGraphQLAPIIdOutput"
Failed to parse Lambda function cloudformation in [redacted]\amplify\backend\function\signupTrigger\signupTrigger-cloudformation-template.json

Error: Could not find ref for "apiGQLAPIGraphQLAPIIdOutput"

Failed to start API Mock endpoint Error: Lambda function createRenderJob does not exist in your project.
Please run amplify add function
```
the functions are definitely in the project and were created by `amplify add function`

and added this @function directive
```
type RenderJobMutation {
  createRenderJobLambda(id:String): String @function(name: "createRenderJob-${env}")
}
```

**Comment by simontegg** [comment]
Having a similar issue. Adding a function to the api breaks the `amplify mock`
```
Failed to parse Lambda function cloudformation in [redacted]/amplify/backend/function/authCreateAuthChallenge/authCreateAuthChallenge-cloudformation-template.json

Error: Could not find ref for "CHALLENGEANSWER"
Failed to parse Lambda function cloudformation in  [redacted]/amplify/backend/function/authPreSignup/authPreSignup-cloudformation-template.json

Error: Could not find ref for "DOMAINWHITELIST"
Failed to parse Lambda function cloudformation in  [redacted]/amplify/backend/function/authPreTokenGeneration/authPreTokenGeneration-cloudformation-template.json

Error: Could not find ref for "relationshipTable"
Failed to parse Lambda function cloudformation in  [redacted]/amplify/backend/function/authVerifyAuthChallengeResponse/authVerifyAuthChallengeResponse-cloudformation-template.json

Error: Could not find ref for "RECAPTCHASECRET"
Failed to parse Lambda function cloudformation in  [redacted]/amplify/backend/function/createOrgTransaction/createOrgTransaction-cloudformation-template.json

Error: Could not find ref for "relationshipTable"
Failed to start API Mock endpoint Error: Lambda function createOrgTransaction does not exist in your project. 
Please run amplify add function
```

I've added all these refs to the cloudformation  and the `parameters.json`.  And it does not work  Though the cloudformation does not appear to depend on them?

**Comment by imgntn** [comment]
hey @simontegg not sure if it will help, but here's what my parameters.json looks like after adding the refs.  made the error go away

```
{
    "permissions": {
        "api": {
            "GQLAPI": [
                "create",
                "read",
                "update",
                "delete"
            ]
        }
    },
    "apiGQLAPIGraphQLAPIEndpointOutput":"apiGQLAPIGraphQLAPIEndpointOutput",
    "apiGQLAPIGraphQLAPIIdOutput":"apiGQLAPIGraphQLAPIIdOutput"
}
```

**Comment by PerrySong** [comment]
Experiencing the same issue,

Error: Could not find ref for:
```
"authshowcasesocial29b118bf2UserPoolId"
"apishowcasesocial2GraphQLAPIIdOutput"
"apishowcasesocial2GraphQLAPIEndpointOutput"
"functionshowcasesocial29848a5e0Name"
```

After manually added it to function-parameters.json, amplify mock api can only give me an empty page with this console error:

```
Failed to load resource: the server responded with a status of 404 (Not Found)
```

**Comment by dtelaroli** [comment]
+1

**Comment by Lasim** [comment]
Well got the same issue with storage.
Can't mock anymore my lambda function.

**Comment by dmsolutionz** [comment]
I was facing the same issue as stated in here.

**Frustrating bits:**
- Other functions were working as expected.
- Ref's not being referenced.

Figuring that I had working functions using `@function(name: "functionName-${env}")` I looked at the `amplify-meta.json` file to seek out the differences, it turns out that if you have your function referencing your API it will force it to not be recognised, this feels like a bug but im not entirely sure.

Basically removing:

![ss4469](https://user-images.githubusercontent.com/12456115/77675727-37757700-6f85-11ea-9419-e42573785347.jpg)

Got the function to show (and work).

**Comment by edeuxk** [comment]
same issue

**Comment by cyrfer** [comment]
@yuth the commits you added don't fix the issue. Please re-open this ticket. I'm using amplify-cli 4.32.0 on OSX High Sierra & WSL2 (windows subsystem for linux: ubuntu).

@mooser this does not work for me:
> You can specify the missing refs in function-parameters.json

I've had no success passing parameters to the function's cloudformation template with `amplify mock`. The only hack I've had success with is to hard code values assigned in env vars section of the function's cloudformation template.
